File: backend/node/api/src/domain/entities/pipeline.ts
import {
	InvalidCanvasError,
	InvalidStateTransitionError,
	StepInconsistencyError,
} from "../errors";
import { UUID } from "../value-objects/uuid";
import { Step } from "./step";

export type StepInput = {
	id: string;
	name?: string;
	description?: string;
	operation: string;
	by: number;
	nextStepId?: string;
};

type CreatePayload = {
	name?: string;
	description?: string;
	initialStepId?: string;
	canvas?: string;
};

type RestorePayload = {
	id: string;
	name?: string;
	description?: string;
	initialStepId?: string;
	steps?: Step[];
	canvas?: string;
	createdAt: Date;
	updatedAt: Date;
};

type ConstructorPayload = {
	id: UUID;
	name?: string;
	description?: string;
	initialStepId?: UUID;
	steps: Step[];
	canvas: string;
	createdAt: Date;
	updatedAt: Date;
};

export class Pipeline {
	private id: UUID;
	private name?: string;
	private description?: string;
	private initialStepId?: UUID;
	private steps: Step[];
	private canvas: string;
	private createdAt: Date;
	private updatedAt: Date;

	constructor(payload: ConstructorPayload) {
		this.id = payload.id;
		this.name = payload.name;
		this.description = payload.description;
		this.initialStepId = payload.initialStepId;
		this.steps = payload.steps || [];
		this.canvas = payload.canvas;
		this.createdAt = payload.createdAt;
		this.updatedAt = payload.updatedAt;
	}

	private static validateCanvas(
		canvas: string,
	): [undefined, undefined] | [undefined, Error] {
		try {
			JSON.parse(canvas);
		} catch {
			return [undefined, new InvalidCanvasError(canvas)];
		}
		return [undefined, undefined];
	}

	static create(
		payload: CreatePayload,
	): [Pipeline, undefined] | [undefined, Error] {
		const now = new Date();
		const id = UUID.create();

		const [stepId, stepIdError] = payload.initialStepId
			? UUID.restore(payload.initialStepId)
			: [undefined, undefined];
		if (!!stepIdError) return [undefined, stepIdError];

		const canvas = payload.canvas || JSON.stringify({});

		const [, canvasError] = Pipeline.validateCanvas(canvas);
		if (canvasError) return [undefined, canvasError];

		const objPayload: ConstructorPayload = {
			...payload,
			id,
			initialStepId: stepId,
			steps: [],
			canvas,
			createdAt: now,
			updatedAt: now,
		};

		return [new Pipeline(objPayload), undefined];
	}

	static restore(
		payload: RestorePayload,
	): [Pipeline, undefined] | [undefined, Error] {
		const [id, idError] = UUID.restore(payload.id);
		const [initialStepId, initialStepIdError] = payload.initialStepId
			? UUID.restore(payload.initialStepId)
			: [undefined, undefined];

		if (!!idError) return [undefined, idError];
		if (!!initialStepIdError) return [undefined, initialStepIdError];

		const canvas = payload.canvas || JSON.stringify({});

		const [, canvasError] = Pipeline.validateCanvas(canvas);
		if (canvasError) return [undefined, canvasError];

		return [
			new Pipeline({
				...payload,
				id,
				initialStepId,
				steps: payload.steps || [],
				canvas,
			}),
			undefined,
		];
	}

	static validateStepChain(
		steps: Step[],
	): [undefined, undefined] | [undefined, Error] {
		for (let i = 0; i < steps.length; i++) {
			const currentStep = steps[i]!;
			const isLastStep = i === steps.length - 1;

			if (isLastStep) {
				if (currentStep.getNextStepId()) {
					return [
						undefined,
						new InvalidStateTransitionError(
							"step",
							"undefined",
							"last step should not have nextStepId",
						),
					];
				}
				continue;
			}

			const nextStep = steps[i + 1]!;

			if (currentStep.getNextStepId() !== nextStep.getId()) {
				return [
					undefined,
					new InvalidStateTransitionError(
						"step",
						currentStep.getNextStepId() || "undefined",
						nextStep.getId(),
					),
				];
			}
		}

		return [undefined, undefined];
	}

	static restoreSteps(
		pipeline: Pipeline,
		stepInputs: StepInput[],
	): [Step[], undefined] | [undefined, Error] {
		const existingStepIds = new Set(pipeline.getSteps().map((s) => s.getId()));
		const chain: Step[] = [];
		const pipelineId = pipeline.getId();

		for (const stepInput of stepInputs) {
			if (existingStepIds.has(stepInput.id)) {
				const existingStep = pipeline
					.getSteps()
					.find((s) => s.getId() === stepInput.id);
				if (!existingStep)
					return [undefined, new StepInconsistencyError(stepInput.id)];

				const [step, restoreError] = Step.restore({
					...stepInput,
					pipelineId,
					createdAt: existingStep.getCreatedAt(),
					updatedAt: new Date(),
				});
				if (restoreError) return [undefined, restoreError];
				chain.push(step);
			} else {
				const [step, createError] = Step.create({
					...stepInput,
					pipelineId,
				});
				if (createError) return [undefined, createError];
				chain.push(step);
			}
		}

		return [chain, undefined];
	}

	setInitialStepId(
		initialStepId?: string,
	): [undefined, undefined] | [undefined, Error] {
		if (initialStepId) {
			const [id, error] = UUID.restore(initialStepId);
			if (error) return [undefined, error];
			this.initialStepId = id;
		} else {
			this.initialStepId = undefined;
		}
		this.updatedAt = new Date();
		return [undefined, undefined];
	}

	setName(name?: string): void {
		this.name = name;
		this.updatedAt = new Date();
	}

	setDescription(description?: string): void {
		this.description = description;
		this.updatedAt = new Date();
	}

	setSteps(steps: Step[]): [undefined, undefined] | [undefined, Error] {
		const [, validationError] = Pipeline.validateStepChain(steps);
		if (validationError) return [undefined, validationError];

		this.steps = steps;
		this.initialStepId = steps[0]
			? UUID.restore(steps[0].getId())[0]
			: undefined;
		this.updatedAt = new Date();
		return [undefined, undefined];
	}

	getId() {
		return this.id.getValue();
	}

	getName() {
		return this.name;
	}

	getDescription() {
		return this.description;
	}

	getInitialStepId() {
		return this.initialStepId?.getValue();
	}

	getCreatedAt() {
		return this.createdAt;
	}

	getUpdatedAt() {
		return this.updatedAt;
	}

	getSteps() {
		return this.steps;
	}

	getCanvas() {
		return this.canvas;
	}

	setCanvas(canvas: string): [undefined, undefined] | [undefined, Error] {
		const [, canvasError] = Pipeline.validateCanvas(canvas);
		if (canvasError) return [undefined, canvasError];

		this.canvas = canvas;
		this.updatedAt = new Date();
		return [undefined, undefined];
	}
}

File: backend/go/api/internal/pipeline/pipeline.go
package pipeline

import (
	"encoding/json"
	"slices"
	"time"

	"github.com/maurodesouza/calculation-pipeline/backend/go/api/internal/shared/vo"
	"github.com/maurodesouza/calculation-pipeline/backend/go/api/utils"
)

type Pipeline struct {
	id            vo.UUID
	name          *string
	description   *string
	initialStepId *vo.UUID
	steps         []Step
	canvas        string
	createdAt     time.Time
	updatedAt     time.Time
}

type NewPipelinePayload struct {
	ID            *string
	Name          *string
	Description   *string
	InitialStepId *string
	Canvas        *string
}

func NewPipeline(payload NewPipelinePayload) (*Pipeline, error) {
	initialStepId, err := getInitialStepID(payload.InitialStepId)
	if err != nil {
		return nil, err
	}
	canvas := getCanvas(payload.Canvas)
	if err := validateCanvas(canvas); err != nil {
		return nil, err
	}

	id := vo.NewUUID()
	now := time.Now()

	return &Pipeline{
		id:            id,
		name:          payload.Name,
		description:   payload.Description,
		initialStepId: initialStepId,
		steps:         []Step{},
		canvas:        canvas,
		createdAt:     now,
		updatedAt:     now,
	}, nil
}

type RestorePipelinePayload struct {
	ID            string
	Name          *string
	Description   *string
	InitialStepId *string
	Steps         []Step
	Canvas        *string
	CreatedAt     string
	UpdatedAt     string
}

func RestorePipeline(payload RestorePipelinePayload) (*Pipeline, error) {
	id, err := vo.RestoreUUID(&payload.ID)
	if err != nil {
		return nil, err
	}

	initialStepId, err := getInitialStepID(payload.InitialStepId)
	if err != nil {
		return nil, err
	}
	canvas := getCanvas(payload.Canvas)
	if err := validateCanvas(canvas); err != nil {
		return nil, err
	}

	createdAt, err := time.Parse(time.RFC3339, payload.CreatedAt)
	if err != nil {
		return nil, err
	}

	updatedAt, err := time.Parse(time.RFC3339, payload.UpdatedAt)
	if err != nil {
		return nil, err
	}

	return &Pipeline{
		id:            id,
		name:          payload.Name,
		description:   payload.Description,
		initialStepId: initialStepId,
		steps:         payload.Steps,
		canvas:        canvas,
		createdAt:     createdAt,
		updatedAt:     updatedAt,
	}, nil
}

type StepInput struct {
	ID          string
	Name        *string
	Description *string
	Operation   string
	By          int
	NextStepId  *string
}

func RestoreSteps(pipeline *Pipeline, stepInputs []StepInput) ([]Step, error) {
	var (
		pipelineId       = pipeline.GetId()
		existingStepsIds []string
		chain            []Step
	)

	existingSteps := pipeline.GetSteps()

	for _, step := range existingSteps {
		existingStepsIds = append(existingStepsIds, step.GetId())
	}

	for _, stepInput := range stepInputs {
		if slices.Contains(existingStepsIds, stepInput.ID) {
			existingStep := existingSteps[slices.IndexFunc(existingSteps, func(step Step) bool {
				return step.GetId() == stepInput.ID
			})]

			step, err := RestoreStep(RestoreStepPayload{
				ID:          stepInput.ID,
				Name:        stepInput.Name,
				Description: stepInput.Description,
				Operation:   stepInput.Operation,
				By:          stepInput.By,
				NextStepID:  stepInput.NextStepId,
				PipelineID:  pipelineId,
				CreatedAt:   existingStep.GetCreatedAt(),
				UpdatedAt:   time.Now().Format(time.RFC3339),
			})

			if err != nil {
				return nil, err
			}

			chain = append(chain, *step)
		} else {
			step, err := NewStep(NewStepPayload{
				ID:          &stepInput.ID,
				Name:        stepInput.Name,
				Description: stepInput.Description,
				Operation:   stepInput.Operation,
				By:          stepInput.By,
				NextStepID:  stepInput.NextStepId,
				PipelineID:  &pipelineId,
			})

			if err != nil {
				return nil, err
			}

			chain = append(chain, *step)
		}
	}

	return chain, nil
}

func (entity *Pipeline) SetName(value *string) {
	entity.name = value
	entity.updatedAt = time.Now()
}

func (entity *Pipeline) SetDescription(value *string) {
	entity.description = value
	entity.updatedAt = time.Now()
}

func (entity *Pipeline) SetCanvas(value *string) error {
	err := validateCanvas(*value)

	if err != nil {
		return err
	}

	entity.canvas = *value
	entity.updatedAt = time.Now()

	return nil
}

func (entity *Pipeline) SetSteps(steps []Step) error {
	err := validateStepChain(steps)
	if err != nil {
		return err
	}

	initialStepId, err := getInitialStepIDFromSteps(steps)
	if err != nil {
		return err
	}

	entity.steps = steps
	entity.initialStepId = initialStepId
	entity.updatedAt = time.Now()

	return nil
}

func (entity *Pipeline) GetId() string {
	return entity.id.GetValue()
}

func (entity *Pipeline) GetName() *string {
	return entity.name
}

func (entity *Pipeline) GetDescription() *string {
	return entity.description
}

func (entity *Pipeline) GetInitialStepId() *string {
	if entity.initialStepId == nil {
		return nil
	}
	value := entity.initialStepId.GetValue()
	return &value
}

func (entity *Pipeline) GetCreatedAt() string {
	return entity.createdAt.Format(time.RFC3339)
}

func (entity *Pipeline) GetUpdatedAt() string {
	return entity.updatedAt.Format(time.RFC3339)
}

func (entity *Pipeline) GetSteps() []Step {
	return entity.steps
}

func (entity *Pipeline) GetCanvas() string {
	return entity.canvas
}

func validateStepChain(steps []Step) error {
	for index, step := range steps {
		isLast := index == len(steps)-1

		if isLast {
			nextStepId := step.GetNextStepId()

			if !utils.IsStringNilOrEmpty(&nextStepId) {
				return InvalidStateTransitionError("step", nextStepId, "last step should not have nextStepId")
			}
		}

		nextStep := steps[index+1]
		nextStepId := step.GetNextStepId()

		if nextStepId != nextStep.GetId() {
			return InvalidStateTransitionError("step", nextStepId, nextStep.GetId())
		}
	}

	return nil
}

func validateCanvas(value string) error {
	var v any

	if err := json.Unmarshal([]byte(value), &v); err != nil {
		return err
	}

	return nil
}

func getInitialStepID(value *string) (*vo.UUID, error) {
	if value == nil {
		return nil, nil
	}

	uuid, err := vo.RestoreUUID(value)
	if err != nil {
		return nil, err
	}

	return &uuid, nil
}

func getInitialStepIDFromSteps(steps []Step) (*vo.UUID, error) {
	if len(steps) == 0 {
		return nil, nil
	}

	id := steps[0].GetId()
	uuid, err := vo.RestoreUUID(&id)

	if err != nil {
		return nil, err
	}

	return &uuid, nil
}

func getCanvas(value *string) string {
	if utils.IsStringNilOrEmpty(value) {
		return "{}"
	}

	return *value
}
